### ISSUE TITLE
Check dqueue_t for overflows

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2021]
See #2020 for details, this is high priority as a widely used data structure and should be done for 0.10.

It is unclear to me why it increases the capacity to `2*old_size + 1` instead of `2*old_size`: 

https://github.com/igraph/igraph/blob/develop/src/core/dqueue.pmt#L301

**Comment by ntamas** [comment]
> It is unclear to me why it increases the capacity to 2*old_size + 1 instead of 2*old_size:

Probably no  particular reason; I'd say change this to `2 * old_size` for sake of consistency and run all unit tests. If they pass, then it's okay. If they don't, then let's figure out why. If the tests pass but there are performance regressions later, iet's worry about it when we see those regressions, and if indeed this is the case then let's document why these regressions are solved by `+ 1`.

**Comment by GroteGnoom** [comment]
Can't its size be zero? That's why I'd usually do 2*x + 1 in these cases.

**Comment by ntamas** [comment]
I think that for _vectors_ we've already ensured that we allocate at least one item if the user requested a size of zero (because `malloc(0)` is undefined behaviour). We could still have `2 * size + 1` for dqueue because its code was copied from vector _before_ the time we've fixed this in vector. Nevertheless, the _capacity_ of the dqueue (i.e. the number of allocated items) should not ever be zero because of said undefined behaviour in `malloc()`.
